File: src/niepce/modules/darkroom/imagecanvas.cpp
/*
 * niepce - modules/darkroom/imagecanvas.cpp
 *
 * Copyright (C) 2008-2022 Hubert Figuière
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <cairomm/context.h>
#include <gdkmm.h>
#include <gtkmm/snapshot.h>

#include "fwk/base/debug.hpp"
#include "fwk/base/geometry.hpp"

#include "imagecanvas.hpp"

namespace dr {

#define IMAGE_INSET 6
#define SHADOW_OFFSET 3


ImageCanvas::ImageCanvas()
    : m_need_redisplay(true),
      m_resized(false),
      m_zoom_mode(ZoomMode::FIT)
{
    set_draw_func([this] (const Cairo::RefPtr<Cairo::Context>& cr, int w, int h) {
        this->on_draw(cr, w, h);
    });
}

void ImageCanvas::set_image(const ncr::Image::Ptr & img)
{
    m_need_redisplay = true;
    m_image_reloaded_cid.disconnect();
    m_image = img;
    m_image_reloaded_cid = m_image->signal_update.connect(
        sigc::mem_fun(*this, &ImageCanvas::on_image_reloaded));
}

void ImageCanvas::on_image_reloaded()
{
    m_need_redisplay = true;
    invalidate();
}

void ImageCanvas::invalidate()
{
    queue_draw();
}

double ImageCanvas::_calc_image_scale(int img_w, int img_h)
{
    double b_w, b_h;
    b_w = get_width() - (IMAGE_INSET * 2);
    b_h = get_height() - (IMAGE_INSET * 2);

    double scale_w = b_w / img_w;
    double scale_h = b_h / img_h;
    return std::min(scale_w, scale_h);
}

void ImageCanvas::_calc_image_frame(int img_w, int img_h,
                                   double & x, double & y,
                                   double & width, double & height)
{
    double b_w, b_h;
    b_w = get_width();
    b_h = get_height();
//    DBG_OUT("bounds %f %f", b_w, b_h);
    x = (b_w - img_w) / 2;
    y = (b_h - img_h) / 2;
    width = img_w;
    height = img_h;
//    DBG_OUT("image frame %f %f %f %f", x, y, width, height);
}

Glib::RefPtr<Gdk::Paintable> ImageCanvas::_get_error_placeholder()
{
    Glib::RefPtr<Gdk::Texture> s;
    try {
        s = Gdk::Texture::create_from_resource(
            "/org/gnome/Niepce/pixmaps/niepce-image-generic.png");
    }
    catch(...) {
    }

    return std::static_pointer_cast<Gdk::Paintable>(s);
}

Glib::RefPtr<Gdk::Paintable> ImageCanvas::_get_missing_placeholder()
{
    Glib::RefPtr<Gdk::Texture> s;
    try {
        s = Gdk::Texture::create_from_resource(
            "/org/gnome/Niepce/pixmaps/niepce-image-missing.png");
    }
    catch(...) {
    }

    return std::static_pointer_cast<Gdk::Paintable>(s);
}

void ImageCanvas::on_resize(int x, int y)
{
    m_resized = true;

    Gtk::DrawingArea::on_resize(x, y);
}

bool ImageCanvas::on_draw(const Cairo::RefPtr<Cairo::Context>& context, int, int)
{
    // no image, just pass.
    if(!m_image) {
        DBG_OUT("no image");
        return false;
    }

    if(m_need_redisplay || m_resized) {
        _redisplay();

        Cairo::RefPtr<Cairo::ImageSurface> img_s;

        int img_w, img_h;
        img_w = img_h = 0;
        double scale = 1.0;

        if (m_image->get_status() < ncr::Image::Status::ERROR) {

            // calculate the image scale
            img_w = m_image->get_original_width();
            img_h = m_image->get_original_height();
            DBG_OUT("image w = %d ; h = %d", img_w, img_h);
            scale = _calc_image_scale(img_w, img_h);
            DBG_OUT("scale = %f", scale);
            m_image->set_output_scale(scale);


            // query the image.
            img_s = m_image->cairo_surface_for_display();
        } else {
            // XXX fix this
            img_w = img_h = 32;
        }

        int canvas_h, canvas_w;
        canvas_h = get_height();
        canvas_w = get_width();
        DBG_OUT("canvas w = %d ; h = %d", canvas_w, canvas_h);

        m_backingstore
            = Cairo::Surface::create(img_s, Cairo::CONTENT_COLOR,
                                     canvas_w, canvas_h);
        Cairo::RefPtr<Cairo::Context> sc
            = Cairo::Context::create(m_backingstore);

//        sc->set_antialias(Cairo::ANTIALIAS_NONE);

        // paint the background
        auto ctxt = get_style_context();
        ctxt->context_save();
        ctxt->set_state(Gtk::StateFlags::NORMAL);
        ctxt->render_background(sc, 0, 0, canvas_w, canvas_h);
        ctxt->context_restore();

        double out_w = (img_w * scale);
        double out_h = (img_h * scale);
        double x = (canvas_w - out_w) / 2;
        double y = (canvas_h - out_h) / 2;
        DBG_OUT("x = %f ; y = %f", x, y);

        sc->rectangle(x + SHADOW_OFFSET, y + SHADOW_OFFSET + 1, out_w, out_h);
        sc->set_source_rgb(0.0, 0.0, 0.0);
        sc->fill();

        if (img_s) {
            sc->set_source(img_s, x, y);
            sc->paint();
        } else  {
            DBG_OUT("no image loaded");
            Glib::RefPtr<Gdk::Paintable> icon;
            if (m_image->get_status() == ncr::Image::Status::NOT_FOUND) {
                icon = _get_missing_placeholder();
            } else {
                icon = _get_error_placeholder();
            }
            img_w = icon->get_intrinsic_width();
            img_h = icon->get_intrinsic_height();
            auto snapshot = Gtk::Snapshot::create();
            icon->snapshot(snapshot, img_w, img_h);
            DBG_ASSERT(!!img_s, "img_s not loaded");
            GskRenderNode* node = gtk_snapshot_to_node(snapshot->gobj());
            auto surface = Cairo::Surface::create(m_backingstore, 0, 0, img_w, img_h);
            gsk_render_node_draw(node, sc->cobj());
        }


//        sc->set_source_rgb(1.0, 1.0, 1.0);
//        sc->set_line_width(1.0);
//        sc->rectangle(x, y, out_w, out_h);
//        sc->stroke();

        m_need_redisplay = false;
        m_resized = false;
    }

    context->set_source(m_backingstore, 0, 0);
    context->paint();

    return true;
}


void ImageCanvas::_redisplay()
{
    if (m_image->get_status() != ncr::Image::Status::LOADED) {
        ERR_OUT("Image is in not loaded - status %d", (int)m_image->get_status());
        return;
    }
    int img_w, img_h;
    img_w = m_image->get_original_width();
    img_h = m_image->get_original_height();
    DBG_OUT("set image w %d h %d", img_w, img_h);

    fwk::Rect dest(0,0, get_width() - 8, get_height() - 8);
    fwk::Rect source(0,0, img_w, img_h);
    fwk::Rect frame;
    switch(m_zoom_mode)
    {
    case ZoomMode::FIT:
        frame = source.fit_into(dest);
        break;
    case ZoomMode::FILL:
        frame = source.fill_into(dest);
        break;
    default:
        frame = source;
        break;
    }
}

}


/*
  Local Variables:
  mode:c++
  c-file-style:"stroustrup"
  c-file-offsets:((innamespace . 0))
  indent-tabs-mode:nil
  fill-column:99
  End:
*/

File: niepce-main/src/modules/darkroom/image_canvas.rs
/*
 * niepce - niepce/modules/darkroom/image_canvas.rs
 *
 * Copyright (C) 2023-2025 Hubert Figuière
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

use gtk4::prelude::*;
use gtk4::subclass::prelude::*;
use npc_fwk::{glib, gtk4};

use npc_fwk::toolkit::ImageBitmap;

#[derive(Clone, Copy, Default, glib::Enum)]
#[enum_type(name = "NcrZoomMode")]
enum ZoomMode {
    None = 0,
    #[default]
    Fit,
    Fill,
    /// 100%
    OneOne,
    Custom, // xxx this should carry a value
}

glib::wrapper! {
    pub struct ImageCanvas(
        ObjectSubclass<imp::ImageCanvas>)
        @extends gtk4::DrawingArea, gtk4::Widget,
        @implements gtk4::Accessible, gtk4::Buildable, gtk4::ConstraintTarget;
}

impl Default for ImageCanvas {
    fn default() -> ImageCanvas {
        ImageCanvas::new()
    }
}

impl ImageCanvas {
    pub fn new() -> ImageCanvas {
        glib::Object::new()
    }

    pub fn set_image(&self, image: ImageBitmap) {
        let imp = self.imp();
        imp.request_redisplay();
        imp.image.replace(Some(image));
        self.queue_draw();
    }

    pub fn set_image_none(&self) {
        self.imp().request_redisplay();
        self.imp().image.replace(None);
        self.queue_draw();
    }
}

mod imp {
    use std::cell::{Cell, RefCell};

    use gtk4::prelude::*;
    use gtk4::subclass::prelude::*;
    use npc_fwk::{cairo, gdk4, glib, graphene, gtk4};

    use super::ZoomMode;
    use npc_fwk::base::Rect;
    use npc_fwk::toolkit::ImageBitmap;
    use npc_fwk::{dbg_out, on_err_out};

    const IMAGE_INSET: f64 = 6.0;
    const SHADOW_OFFSET: f64 = 3.0;

    lazy_static::lazy_static! {
        static ref ERROR_PLACEHOLDER: gdk4::Texture = gdk4::Texture::from_resource("/net/figuiere/Niepce/pixmaps/niepce-image-generic.png");
        static ref MISSING_PLACEHOLDER: gdk4::Texture = gdk4::Texture::from_resource("/net/figuiere/Niepce/pixmaps/niepce-image-missing.png");
    }

    #[derive(Default)]
    pub struct ImageCanvas {
        need_redisplay: Cell<bool>,
        resized: Cell<bool>,
        zoom_mode: ZoomMode,
        pub(super) image: RefCell<Option<ImageBitmap>>,
    }

    impl ImageCanvas {
        pub(super) fn request_redisplay(&self) {
            self.need_redisplay.set(true);
        }

        fn calc_image_scale(&self, img_w: u32, img_h: u32) -> f64 {
            let obj = self.obj();
            let b_w = obj.width() as f64 - (IMAGE_INSET * 2.0);
            let b_h = obj.height() as f64 - (IMAGE_INSET * 2.0);

            let scale_w = b_w / img_w as f64;
            let scale_h = b_h / img_h as f64;

            scale_w.min(scale_h)
        }

        fn on_draw(this: &gtk4::DrawingArea, ctx: &cairo::Context, w: i32, h: i32) {
            if let Some(this) = this.downcast_ref::<super::ImageCanvas>() {
                this.imp().on_draw_(ctx, w, h);
            }
        }

        fn on_draw_(&self, context: &cairo::Context, _: i32, _: i32) {
            if self.need_redisplay.get() || self.resized.get() {
                self.redisplay();

                let mut img_w = ERROR_PLACEHOLDER.width() as u32;
                let mut img_h = ERROR_PLACEHOLDER.height() as u32;
                let texture = {
                    self.image
                        .borrow()
                        .as_ref()
                        .map(|image| {
                            img_w = image.original_width();
                            img_h = image.original_height();
                            // query the image.
                            image.to_gdk_texture()
                        })
                        .or_else(|| Some(MISSING_PLACEHOLDER.clone()))
                };

                dbg_out!("texture? {}", texture.is_some());

                let obj = self.obj();
                let canvas_h = obj.height();
                let canvas_w = obj.width();
                dbg_out!("canvas w = {canvas_w} ; h = {canvas_h}");

                dbg_out!("image w = {img_w} ; h = {img_h}");
                let scale = self.calc_image_scale(img_w, img_h);
                dbg_out!("scale = {scale}");

                let out_w = img_w as f64 * scale;
                let out_h = img_h as f64 * scale;
                let x = (canvas_w as f64 - out_w) / 2.0;
                let y = (canvas_h as f64 - out_h) / 2.0;
                dbg_out!("x = {x} ; y = {y}");

                context.rectangle(x + SHADOW_OFFSET, y + SHADOW_OFFSET + 1.0, out_w, out_h);
                context.set_source_rgb(0.0, 0.0, 0.0);
                on_err_out!(context.fill());

                if let Some(ref texture) = texture {
                    let snapshot = gtk4::Snapshot::new();
                    snapshot.translate(&graphene::Point::new(x as f32, y as f32));
                    texture.snapshot(&snapshot, out_w, out_h);
                    if let Some(node) = snapshot.to_node() {
                        node.draw(context);
                    }
                }
            }

            self.need_redisplay.set(false);
            self.resized.set(false);
        }

        /// Recalculate the display frame.
        fn redisplay(&self) -> Option<Rect> {
            if let Some(image) = self.image.borrow().as_ref() {
                let img_w = image.original_width();
                let img_h = image.original_height();
                dbg_out!("set image w {img_w} h {img_h}");

                let obj = self.obj();
                let dest = Rect::new(0, 0, (obj.width() - 8) as u32, (obj.height() - 8) as u32);
                let source = Rect::new(0, 0, img_w, img_h);
                let frame = match self.zoom_mode {
                    ZoomMode::Fit => source.fit_into(&dest),
                    ZoomMode::Fill => source.fill_into(&dest),
                    _ => source,
                };

                Some(frame)
            } else {
                None
            }
        }
    }

    #[glib::object_subclass]
    impl ObjectSubclass for ImageCanvas {
        const NAME: &'static str = "NcrImageCanvas";
        type Type = super::ImageCanvas;
        type ParentType = gtk4::DrawingArea;
    }

    impl ObjectImpl for ImageCanvas {
        fn constructed(&self) {
            self.parent_constructed();

            let obj = self.obj();
            obj.set_draw_func(Self::on_draw);
            obj.connect_resize(|this, _, _| {
                this.imp().resized.set(true);
            });
        }
    }

    impl DrawingAreaImpl for ImageCanvas {}

    impl WidgetImpl for ImageCanvas {}
}
